commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -248,23 +248,26 @@ final class JsonReader(val json: String) {
       advance()
     } else if (isNextDigit) {
       parseDigits()
-    } else throw new ReadFailure("Expected '-' or digit")
+    } else throw new ReadFailure(s"Expected '-' or digit, got ${json.charAt(i)}")
 
     if (isNext('.')) {
-      decimal = true
       advance()
       parseDigits()
-      if (isNext('e') || isNext('E')) {
+    }
+
+    //Double.MinPositiveValue.toString is 5e-324 in JS. Written by scientists, for scientists.
+    if (isNext('e') || isNext('E')) {
+      advance()
+      if (isNext('-') || isNext('+')) {
         advance()
-        if (isNext('-') || isNext('+')) {
-          advance()
-          parseDigits()
-        } else throw new ReadFailure("Expected '+' or '-'")
       }
+      if (isNextDigit) {
+        parseDigits()
+      } else throw new ReadFailure(s"Expected a digit, got ${json.charAt(i)}")
     }
 
     val str = json.substring(start, i)
-    if (decimal) str.toDouble else str.toInt
+    str
```

Comment:
The problem here is that `readString()` will now work even if there was a number.

Reply:
Can you propose an alternative? We're omitting so many other checks for performance reasons and the fact, that the usage pattern explicitly specifies the return type.

Reply:
Keep a flag along raw value or something. We have benchmarks so we can see if it breaks performance.